test/engine/engine_support_test.cc

```diff
@@ -37,6 +37,121 @@ using ::testing::ContainsRegex;
 using ::testing::MatchesRegex;
 using ::testing::Pointwise;
 using ::testing::ElementsAreArray;
+using AngMomMatTest = MujocoTest;
+
+static constexpr char AngMomTestingModel[] = R"(
+<mujoco>
+<option integrator = "RK4" timestep = "0.001" jacobian = "dense"/>
+<worldbody>
+  <body name="link1">
+    <joint type="hinge" pos="0 0 0" axis="0 1 0"/>
+    <geom type="capsule" size="0.05" fromto="0 0 0  1 0 0"/>
+    <body name="link2" pos="1 0 0">
+      <joint type="ball" />
+      <geom type="capsule" size="0.05" fromto="0 0 0  0 1 0"/>
+    </body>      
+  </body>
+</worldbody>
+</mujoco>
+  )";
+
+// compare subtree angular momentum computed in two ways
+TEST_F(AngMomMatTest, CompareAngMom) {
+  mjModel* model = LoadModelFromString(AngMomTestingModel);
+  int nv = model->nv;
+  int bodyid = mj_name2id(model, mjOBJ_BODY, "link1");
+  mjData* data = mj_makeData(model);
+
+  // let the mechanism move and generate some angular momentum
+  for (int i=0; i < 500; i++) {
+    mj_step(model, data);
+  }
+
+  // get the reference value of angular momentum
+  mj_subtreeVel(model,data);
+  mjtNum angmom_ref[3];
+  mju_copy3(angmom_ref, data->subtree_angmom+3*bodyid);
+
+  // compute angular momentum using the angular momentum matrix
+  mjtNum* angmom_mat = (mjtNum*) mju_malloc(sizeof(mjtNum)*3*nv);
+  mj_subtreeAngMomMat(model, data, angmom_mat, bodyid);
+  mjtNum angmom_test[3];
+  mju_mulMatVec(angmom_test, angmom_mat, data->qvel, 3, nv);
+
+  // compare the two angular momentum values
+  static const mjtNum tol = 1e-3;
+  for(int i=0; i<3; i++) {
+    EXPECT_THAT(angmom_ref[i], DoubleNear(angmom_test[i], tol));
+  }
+
+  mju_free(angmom_mat);
+  mj_deleteData(data);
+  mj_deleteModel(model);
+}
+
+// compare subtree angular momentum matrix computed in two ways: analytical and fdm
+TEST_F(AngMomMatTest, CompareAngMomMats) {
+  mjModel* model = LoadModelFromString(AngMomTestingModel);
+  int nv = model->nv;
+  int bodyid = mj_name2id(model, mjOBJ_BODY, "link1");
+  mjData* data = mj_makeData(model);
+  mjtNum* angmom_mat = (mjtNum*) mju_malloc(sizeof(mjtNum)*3*nv);
+  mjtNum* angmom_mat_fdm = (mjtNum*) mju_malloc(sizeof(mjtNum)*3*nv);
+
+  // let the mechanism move and generate some angular momentum
+  for (int i=0; i < 500; i++) {
+    mj_step(model, data);
+  }
+
+  // compute the angular momentum matrix using the analytical method
+  mj_subtreeAngMomMat(model, data, angmom_mat, bodyid);
+
+  // compute the angular momentum matrix using finite differences
+  static const mjtNum eps = 1e-3;
```

Comment:
This is large. Use eps = 1e-6 and then reduce tol until the test fails, then increase it by one click (factor of 10)

Reply:
Computations using the forward difference method were not converging for different eps values, hence, tried out to the central difference method. Now the analytical and finite difference method results agree to a great extent.